src/internal_events/http.rs

```diff
@@ -42,3 +42,40 @@ impl<'a> InternalEvent for HTTPBadRequest<'a> {
         counter!("http_bad_requests_total", 1);
     }
 }
+
+#[derive(Debug)]
+pub struct HTTPEventMissingMessage;
+
+impl InternalEvent for HTTPEventMissingMessage {
+    fn emit_logs(&self) {
+        warn!(
+            message = "Event missing the message key; dropping event.",
+            rate_limit_secs = 30,
+        );
+    }
+
+    fn emit_metrics(&self) {
+        counter!("events_discarded_total", 1);
+    }
+}
+
+#[derive(Debug)]
+pub struct HTTPEventsSent {
+    pub events_count: usize,
+    pub byte_size: usize,
```

Comment:
This seems to be the source of most of the complications, and I think we could potentially just get rid of it? There will be situations like HTTP streaming bodies where we can't really get it anyway, so maybe simpler to just leave it out and get byte information from other instrumentation like batching. Then hopefully this is a bit simpler and doesn't require things to be passed around in the same way.

Reply:
Maybe do it like this?

https://github.com/timberio/vector/blob/f9bf655dd7aa7a3d08be104e67a9ed26424eeb22/src/http.rs#L119-L136

Reply:
True, but I worry about how we would represent the uncertainty there. If we just use those numbers naively, the resulting metrics would likely be "inaccurate".

Reply:
I would prefer if we avoid capturing the byte sizes of request/response unless it is made available to us by the underlying layers. Though realistically, we can't expect those to be available - nobody does that. I guess we should not pass the bytes here and remove them from the HTTP client logging. The way they're obtained is not intended for how we try to use it.

Reply:
If we remove `byte_size` we still need to find some way to emitting events when they are sent. Or we can emit an event when we encode data and put it in the queue?

Reply:
Assuming we emit an event at the HTTP client - I think we're covered, and there's no need to emit elsewhere.